### PR TITLE
Block editor: try direct drag (outside text editable)

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

direct-drag

## Why?

Because it's more intuitive.

## How?

Enable built-in drag support and set the correct data.

We might want to use the current drag chip, instead of the default preview that the browser provides, not sure. 

## Testing Instructions

Drag an image by click and holding it directly instead of using the toolbar.

### CODE REVIEW

**Comment by youknowriad** [comment]
In my testing this doesn't work in Safari for the moment. It works on Chrome.

A couple of notes:

 - Indeed, the clone is too big, I feel the drag handle as a draggable element might feel better. We can probably have a small "preview" or a small clone exploration too but maybe fine to start with a similar draggable element to the one used when you use the drag handle.
 - With this PR, we lose the possibility to multi-select blocks starting from an image block. I think this is probably fine because for multi-selection, starting from text feels more natural. But I wanted to note it in case we want to account for it somehow.

**Comment by youknowriad** [comment]
Another thing I noticed is that for  image block placeholders", it only works after you select the block. It would be better if we can trigger it without having to select the block first.

**Comment by ellatrix** [comment]
@youknowriad Fixed the issue in Safari, which also fixes the placeholder issue I think.

**Comment by youknowriad** [comment]
Wow, this is feeling so great :)

File: packages/block-editor/src/components/block-draggable/content.scss [hunk 1]
```diff
@@ -18,3 +18,7 @@
 		content: none !important;
 	}
 }
+
+img:not([draggable]) {
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-draggable/content.scss]: What is this about?

File: packages/block-editor/src/components/block-list/use-block-props/index.js [hunk 2]
```diff
@@ -152,6 +152,7 @@ export function useBlockProps( props = {}, { __unstableIsHtml } = {} ) {
 
 	return {
 		tabIndex: blockEditingMode === 'disabled' ? -1 : 0,
+		draggable: true,
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/use-block-props/index.js]: This feels like an impactful change no? Do we expect anything weird?

File: packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js [hunk 3]
```diff
@@ -76,7 +76,34 @@ export function useEventHandlers( { clientId, isSelected } ) {
...
+				if (
+					node !== event.target ||
+					( node.contains( anchorNode ) &&
+						anchorNode.isContentEditable ) ||
+					( node.contains( focusNode ) &&
+						focusNode.isContentEditable )
+				) {
+					event.preventDefault();
+					return;
+				}
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js:94]: Would you mind adding some comments here to clarify what the code does, I think you try to guess the cases where you want to bail early and disable the bahavior but it would be cool to explain the cases.

**Review by youknowriad — COMMENTED** [review]
I noticed that when you have a multi-selection and start dragging, it only drags the image that you clicked on and not the full multi-selection. 

I'm happy for this to be considered a follow-up but here are my expectations for multi-selected blocks:

 - If I click on a block (regardless of the block itself, text or not), Ideally the whole multi-selection moves.
 - If I click on an image outside the multi selection, only that image is dragged.

**Comment by youknowriad** [comment]
Do you think this drag/drop behavior is e2e testable. It would be good to solidify it.

**Comment by draganescu** [comment]
This is grea!!

Sometimes it's very hard to figure out what is going on. In the vide below I try to move things between columns and it's unclear why eventually works and then does not work again.


https://github.com/user-attachments/assets/8cfc7b71-535a-41cf-8992-8844240ce825




Also images can be dragged and dropped on top of other images and nothing happens.

Also:

- maybe we should stick to those chips we have now (not a shaddow of the dragged dom)
- I know @ellatrix dislikes too many experiements but this reaaaly looks like a great experiment to polish while in trunk

**Comment by youknowriad** [comment]
@draganescu Given how this is implemented, my guess is that all of the "dropping" issues you're having in your videos have nothing to do with this PR and may already be present in trunk. Did you try there using the chip?

**Comment by ellatrix** [comment]
Yes, so I would compare what happens when you drag using the block toolbar, it may be that nothing happens then too. The drag transfer from this should be 100% the same, so would be good to know.

File: packages/block-editor/src/components/block-list/use-block-props/index.js (hunk 2 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/use-block-props/index.js]: Yes, because I generalised it to all blocks if there is no text selection. I will polish the conditions a bit and test with other blocks, but it would be cool if you can drag any block. I really love that you can now also drag list items from the bullet etc.

File: packages/block-editor/src/components/block-draggable/content.scss (hunk 1 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/content.scss]: I'll comment, but it's because images are by default draggable (even without the html attribute), so we want to prevent that so that you drag the whole image block instead.

**Comment by ellatrix** [comment]
> I noticed that when you have a multi-selection and start dragging, it only drags the image that you clicked on and not the full multi-selection.

I had not considered multi block selection yet. I might disable that for now

File: packages/block-editor/src/components/block-draggable/content.scss (hunk 1 above)
Comment (youknowriad) [reply packages/block-editor/src/components/block-draggable/content.scss]: Do we want to be more specific though, like target only images that are block wrappers or things like that or do we want this for all images.

If it's for all images, maybe we can check whether it's the only html element with this behavior and whether we should have the same thing for other html elements.

**Comment by draganescu** [comment]
There are issues not caused by the implementation but by the new UX

- in trunk one cannot drag images in content locked patterns
- dragging by the toolbar is certain you drag the selected block. Here I think I am able to drag the column by clicking on the image

**Comment by ellatrix** [comment]
> dragging by the toolbar is certain you drag the selected block. Here I think I am able to drag the column by clicking on the image

Can you show it? What do you have selected?

**Comment by draganescu** [comment]
> Can you show it? What do you have selected?

both are in the video above.

**Comment by ellatrix** [comment]
@draganescu I don't see where it's dragging a whole column? The column that the image was in remained in place?

File: packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js:94]: This has changed now.

File: packages/block-editor/src/components/block-draggable/content.scss (hunk 1 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/content.scss]: Yeah, it was just a quick hack for now, I'll come back to this later

**Comment by ellatrix** [comment]
> Indeed, the clone is too big, I feel the drag handle as a draggable element might feel better. We can probably have a small "preview" or a small clone exploration too but maybe fine to start with a similar draggable element to the one used when you use the drag handle.

This is fixed, changed it to the drag chip.

File: packages/block-editor/src/components/block-draggable/content.scss [hunk 4]
```diff
@@ -5,10 +5,6 @@
 	opacity: 0.05 !important;
 	border-radius: $radius-small !important;
 
-	// Disabling pointer events during the drag event is necessary,
-	// lest the block might affect your drag operation.
-	pointer-events: none !important;
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-draggable/content.scss:10]: @jasmussen You added this a while back in #33950, but not sure why. I'm guessing to not allow dragging within the dragged block's inner blocks. I've fixed this in the PR.

Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/content.scss:10]: We can't do this anymore because pointer-events will end drag immediately.

**Comment by draganescu** [comment]
## More testing:

https://github.com/user-attachments/assets/b69ec18e-e480-4a99-906f-6f037d6d5416

In the video:

- I can drag things that should not be draggable and which dont have the dragging handle in the toolbar (locked things)
- there is a. weird outline in firefox showing for some things being dragged
- there is some ghost elements that show up  (see when I move text from under the image)
- sometimes i can't select text by clicking and dragging
- sometimes clicking in rich text to place caret doesn't work

**Comment by ellatrix** [comment]
@draganescu Yeah, there's a few issues in FF I need to fix

File: packages/block-editor/src/components/use-block-drop-zone/index.js [hunk 5]
```diff
@@ -358,13 +359,29 @@ export default function useBlockDropZone( {
...
+
+				// Check if the target is within any of the dragged blocks.
+				const isTargetWithinDraggedBlocks = draggedBlockClientIds.some(
+					( clientId ) => targetParents.includes( clientId )
+				);
+
+				if ( isTargetWithinDraggedBlocks ) {
+					return;
+				}
+
```
Comment (youknowriad) [inline packages/block-editor/src/components/use-block-drop-zone/index.js:377]: Can should explain a bit why you had to change the drop zone behavior specifically for this PR?

File: packages/components/src/resizable-box/index.tsx [hunk 6]
```diff
@@ -112,6 +112,12 @@ function UnforwardedResizableBox(
 				showHandle && 'has-show-handle',
 				className
 			) }
+			handleComponent={ Object.fromEntries(
+				Object.keys( HANDLE_CLASSES ).map( ( key ) => [
+					key,
+					<div key={ key } tabIndex={ -1 } />,
+				] )
+			) }
```
Comment (youknowriad) [inline packages/components/src/resizable-box/index.tsx:124]: How is this related to the PR?

Comment (ellatrix) [reply packages/components/src/resizable-box/index.tsx:124]: I wouldn't do this if it was't. 😃 Unfortunately `re-resizable` doesn't create proper focusable element for the handles, so the focus goes to the block wrapper instead when you drag them, which triggers block drag.

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:377]: Yes — I removed the `pointer-event:none` for blocks that are being dragged, because it results in drag immediately ending. As a result you can drag the block into the block inner block region. This change prevents inner block regions to be considered as drag targets.

Comment (ellatrix) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:377]: See https://github.com/WordPress/gutenberg/pull/67305#discussion_r1860918190

File: packages/components/src/resizable-box/index.tsx (hunk 6 above)
Comment (ellatrix) [reply packages/components/src/resizable-box/index.tsx:124]: Added an inline comment

**Comment by ellatrix** [comment]
This should be fixed btw

> sometimes i can't select text by clicking and dragging
> sometimes clicking in rich text to place caret doesn't work

**Comment by ellatrix** [comment]
> there is a. weird outline in firefox showing for some things being dragged

Fixed

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 5 above)
Comment (youknowriad) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:377]: Do you think it would be simpler to enable `useDisabled()` in the props of the block when it starts dragging?

File: packages/components/src/resizable-box/index.tsx (hunk 6 above)
Comment (youknowriad) [reply packages/components/src/resizable-box/index.tsx:124]: I'm wondering about `tabIndex={ -1 }`. This seems like it will remove the handles from the tab sequence, which if I'm not wrong is something we want to keep (for example in the handles of the site editor frame resizer and maybe others)

**Comment by ellatrix** [comment]
> I can drag things that should not be draggable and which dont have the dragging handle in the toolbar (locked things)

Fixed too

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:377]: What would that give us? We'd still need this code because if you disable any children in the block, you'll still be able to drag things _in between_ the inner blocks.

Comment (ellatrix) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:377]: This seems like a robust way to prevent allowing dragging blocks within itself, regardless of HTML or CSS

Comment (youknowriad) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:377]: `useDisabled` already exists and disables everything including things that are outside the inner blocks.

Comment (youknowriad) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:377]: The goal is to remove this code entirely, if we can't, it's a bad idea.

File: packages/components/src/resizable-box/index.tsx (hunk 6 above)
Comment (ellatrix) [reply packages/components/src/resizable-box/index.tsx:124]: > This seems like it will remove the handles from the tab sequence

There is no tab sequence, these handles are currently completely unfocusable, that's my point :)

File: packages/block-editor/src/components/block-draggable/content.scss (hunk 4 above)
Comment (jasmussen) [reply packages/block-editor/src/components/block-draggable/content.scss:10]: When you drag a block from A to B, you don't actually _move that block_. A copy of the block is made, and shown near your cursor, the block you're intending to move becomes a "gray slot", and only when you release the mouse button, is the actual block moved, and the copy you were dragging destroyed.

I'm pretty sure that's the reason I set pointer-events to _none_ in that PR: if the block you are moving has other inner blocks, or other interactive elements (I guess ondragover related ones), then even though it shows up as a gray slot, it's still I guess a bit interactive.

If you've solved that, then certainly go forward. Pointer events is always a poor CSS propety to tinker with.

File: packages/components/src/resizable-box/index.tsx (hunk 6 above)
Comment (youknowriad) [reply packages/components/src/resizable-box/index.tsx:124]: The site editor one is within the tab sequence (in trunk) but I confirmed that it's still within the tab sequence after this change.

**Comment by youknowriad** [comment]
In firefox, I can still reproduce the bug where I can't drag a placeholder image block if it's unselected.

**Comment by youknowriad** [comment]
In firefox, I can drag list items (on the dots) but not on Safari.

**Comment by ellatrix** [comment]
@youknowriad Yes, that's because list items in Safari work differently unfortunately. The list item does not include the bullet. Which is different from Chrome and Firefox... Maybe we can have a look at that separately.

Screenshot 2024-11-28 at 09 20 12

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:377]: Yeah, we can't

File: packages/block-editor/src/components/block-draggable/content.scss (hunk 4 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/content.scss:10]: > has other inner blocks, or other interactive elements

Yes, I added a different solution for inner blocks. Also left it for frames, which would completely capture pointer event into another document.

File: packages/block-editor/src/components/block-draggable/content.scss (hunk 1 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/content.scss]: Looked for another solution but this seems to be the only way. So I changed it to only affect images in blocks. Also there doesn't seem to be any other element like this. Tested videos for example and they don't drag.

**Comment by ellatrix** [comment]
> In firefox, I can still reproduce the bug where I can't drag a placeholder image block if it's unselected.

Fixed

File: packages/block-editor/src/components/block-draggable/content.scss (hunk 1 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/content.scss]: Well, it turns out SVGs in Firefox were also draggable: https://github.com/WordPress/gutenberg/pull/67305#issuecomment-2505484832

File: packages/components/src/resizable-box/index.tsx (hunk 6 above)
Comment (ellatrix) [reply packages/components/src/resizable-box/index.tsx:124]: Ah, yes, it seems to have its own `handleComponent` override

File: packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js [hunk 7]
```diff
@@ -21,12 +24,16 @@ import { unlock } from '../../../lock-unlock';
  * @param {string} clientId Block client ID.
  */
 export function useEventHandlers( { clientId, isSelected } ) {
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js:26]: This hook is becoming a bit of a dump every even hook. We should consider a better organization here (maybe a follow-up)

File: packages/block-editor/src/components/rich-text/index.js [hunk 8]
```diff
@@ -431,6 +431,10 @@ export function RichTextWrapper(
 				aria-multiline={ ! disableLineBreaks }
 				aria-readonly={ shouldDisableEditing }
 				{ ...props }
+				// Unset draggable for contentEditable elements because it will
+				// interfere with multi block selection when the contentEditable
+				// and draggable elements are the same element.
+				draggable={ undefined }
```
Comment (youknowriad) [inline packages/block-editor/src/components/rich-text/index.js:438]: Which draggable we're unsetting, the one set by the block wrapper I guess. Maybe it's a good thing to mention.

**Review by youknowriad — COMMENTED** [review]
Let's add some e2e tests and ship.

**Comment by ellatrix** [comment]
Seems there's still a few edge cases in FF to sort out.

**Comment by ellatrix** [comment]
@youknowriad Added an e2e test. I think at some point we should reorganise the and add more systematically. The basic operations are the same as dragging from the toolbar handle.

File: packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js (hunk 7 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js:26]: I don't think it's too bad right now, the hook handles keydown and drag so far, but sure we can split it.

File: packages/block-editor/src/components/rich-text/index.js (hunk 8 above)
Comment (ellatrix) [reply packages/block-editor/src/components/rich-text/index.js:438]: Added a mention

File: packages/block-editor/src/components/block-list/use-block-props/use-firefox-draggable-compatibility.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+/**
+ * WordPress dependencies
+ */
+import { useRefEffect } from '@wordpress/compose';
+
+export function useFirefoxDraggableCompatibility() {
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/use-block-props/use-firefox-draggable-compatibility.js:14]: Can you add a comment explaining what this fixes?

File: packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js [hunk 10]
```diff
@@ -76,7 +83,102 @@ export function useEventHandlers( { clientId, isSelected } ) {
...
+						'is-dragging-components-draggable'
+					);
+					ownerDocument.documentElement.classList.remove(
+						'is-dragging'
+					);
+				}
+
+				ownerDocument.addEventListener( 'dragover', over );
+				ownerDocument.addEventListener( 'dragend', end );
+				ownerDocument.addEventListener( 'drop', end );
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js:174]: It's not really clear to me why we have all these events here.
The other thing is I wonder if some of this code can be shared between the drag handle in the toolbar and here. like a `useBlockDragClone` or things like that. I don't know enough to tell what's the right abstraction though.

File: test/e2e/specs/editor/blocks/spacer.spec.js [hunk 11]
```diff
@@ -43,7 +43,9 @@ test.describe( 'Spacer', () => {
 		expect( await editor.getEditedPostContent() ).toMatchSnapshot();
 
 		await expect(
-			editor.canvas.locator( 'role=document[name="Block: Spacer"i]' )
+			editor.canvas.locator(
+				'role=document[name="Block: Spacer"i] >> css=.components-resizable-box__handle >> [tabindex]'
```
Comment (youknowriad) [inline test/e2e/specs/editor/blocks/spacer.spec.js:47]: This is causing a small visual regression. When you resize the spacer, it kind of loses its nice block outline at the moment.

Comment (youknowriad) [reply test/e2e/specs/editor/blocks/spacer.spec.js:47]: How important it is to make the drag handles focusable?

File: packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js (hunk 10 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js:174]: I plan to refactor all that after this. The block drag component can just be a hook. There's some differences that we can't do here such as rendering the drag chip in place. I rendered it a separate React root for this hook, we should do the same in the other place cause it's much simpler. But also, if we want to use block previews at some point the approach will be very different again.

File: test/e2e/specs/editor/blocks/spacer.spec.js (hunk 11 above)
Comment (ellatrix) [reply test/e2e/specs/editor/blocks/spacer.spec.js:47]: It's important because otherwise we can't differentiate between dragging a block or dragging something else in the block. Block drag (rightly, I think) assumes that if something else within the block is draggable, it also focussed instead of the entire block being focussed, and block drag will ignore it.

Comment (youknowriad) [reply test/e2e/specs/editor/blocks/spacer.spec.js:47]: Can we like check that the currently dragged element is the block wrapper or something instead?

Comment (ellatrix) [reply test/e2e/specs/editor/blocks/spacer.spec.js:47]: The resize handle is in the block wrapper. What can we check?

File: packages/block-editor/src/components/block-list/use-block-props/use-firefox-draggable-compatibility.js (hunk 9 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/use-block-props/use-firefox-draggable-compatibility.js:14]: Added in 76f61df275 and rebased this PR too to get the latest drag and drop fixes.

File: test/e2e/specs/editor/blocks/spacer.spec.js (hunk 11 above)
Comment (youknowriad) [reply test/e2e/specs/editor/blocks/spacer.spec.js:47]: I mean if the element that triggered the drag event is the block wrapper itself.

Comment (ellatrix) [reply test/e2e/specs/editor/blocks/spacer.spec.js:47]: I'm not sure what you mean. But we can't just limit it the block wrapper, otherwise you can drag for example from the image element.

**Review by draganescu — APPROVED** [review]
This works so well! Outstanding 👏🏻

File: test/e2e/specs/editor/blocks/spacer.spec.js (hunk 11 above)
Comment (ellatrix) [reply test/e2e/specs/editor/blocks/spacer.spec.js:47]: > When you resize the spacer, it kind of loses its nice block outline at the moment.

Maybe we can use the block focus outline whenever you focus any element within the block that is not an inner blocks. But idk, what you say here is not really a visual regression imo but rather a fix. For any other block, if you move focus to a button within or some other input field, it removes the outline...

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-editor/src/components/use-block-drop-zone/index.js:377]: Btw, I'm actually seeing the same issue in trunk, so this fixes an additional issue.

Screenshot 2024-11-29 at 18 56 47

**Comment by ellatrix** [comment]
Re-tested Chrome, Safari and Firefox too. Let's merge. I'm sure issues will pop up, but I'm confident that we can fix them quickly.

File: test/e2e/specs/editor/blocks/spacer.spec.js (hunk 11 above)
Comment (youknowriad) [reply test/e2e/specs/editor/blocks/spacer.spec.js:47]: Yes, but the issue is that before this PR the focus stayed on the wrapper which gave it a really nice touch.
Maybe, we can have a special case for the resizable blocks (which could be most blocks in the end), when the resizable handles are focused, the block outlines should be visible as well.

**Comment by youknowriad** [comment]
Thanks for your work on this. This is really great.

**Comment by talldan** [comment]
@ellatrix It's really great to see this shipped 👏 

I think some non-text blocks might need to opt out of this behavior. For example [this Sketch block](https://wordpress.org/plugins/sketch/) can't be used anymore because the block is dragged when a user is attempting to draw something.

It's probably a very small proportion of blocks that will have this kind of issue. I'm not sure if there would be any reliable way to automatically detect them the way contenteditable is being detected.

**Comment by ellatrix** [comment]
@talldan that is usually because the block is doing something wrong. The block will only be dragged if focus is on the outer block wrapper, so if a block doesn't appropriately work and is accessible, it will be dragged yes. So you can disable dragging simply by ensuring the wrapper is not focused 🙂
There's potentially some ways to be more forgiving. I'll check out the block in a bit

**Comment by coder-karen** [comment]
👋 It looks like this PR also affects the Jetpack Tiled Gallery block ([issue](https://github.com/Automattic/jetpack/issues/40747)), though the way the Tiled Gallery block handles interactions isn't perfect, and a solution can be added in Jetpack itself (I've created a PR to prevent the selection issue for now). Still, I wanted to flag it just as another example of a block that's affected in some form, in case it's helpful.
